File: todo/app/src/main/java/com/k/todo/base/BaseFragment.java
package com.k.todo.base;


import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.k.todo.provider.DataSourceListener;
import com.k.todo.provider.MusicProvider;

public class BaseFragment extends Fragment implements DataProvider, DataSourceListener {
    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
    }


    @Override
    public void onLowMemory() {
        super.onLowMemory();
    }

    @Override
    public MusicProvider getMusicProvider() {
        if (getActivity() != null && getActivity() instanceof DataProvider) {
            return ((DataProvider) getActivity()).getMusicProvider();
        }
        return MusicProvider.create();
    }

    @Override
    public void onResume() {
        super.onResume();
        getMusicProvider().registerDataSourceListener(this);
        getMusicProvider().loadData();
    }

    @Override
    public void onPause() {
        super.onPause();
        getMusicProvider().unregisterDataSourceListener(this);
    }

    @Override
    public void onLoaded(Object datas) {

    }

    @Override
    public void onLoadError(Object error) {

    }

    @Override
    public void onDataChanged(Object datas) {

    }

    @Override
    public void onLoading() {

    }
}

File: todo/app/src/main/java/com/k/todo/base/BaseFragment.kt
package com.k.todo.base


import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup

import com.k.todo.provider.DataSourceListener
import com.k.todo.provider.MusicProvider

open class BaseFragment : Fragment(), DataProvider, DataSourceListener {

    override val musicProvider: MusicProvider
        get() {
            var provider: MusicProvider? = null
            if (activity != null && activity is DataProvider) {
                provider = (activity as DataProvider).musicProvider
            }
            return if (provider == null) MusicProvider.create() else provider
        }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        musicProvider.registerDataSourceListener(this)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        super.onCreateOptionsMenu(menu, inflater)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        musicProvider.loadData()
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
    }

    override fun onDetach() {
        super.onDetach()
    }

    override fun onDestroyView() {
        super.onDestroyView()
    }

    override fun onDestroy() {
        super.onDestroy()
        musicProvider.unregisterDataSourceListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        return super.onOptionsItemSelected(item)
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        super.onConfigurationChanged(newConfig)
    }


    override fun onLowMemory() {
        super.onLowMemory()
    }

    override fun onResume() {
        super.onResume()

    }

    override fun onPause() {
        super.onPause()
    }

    override fun onLoaded(datas: Any) {

    }

    override fun onLoadError(error: Any) {

    }

    override fun onDataChanged(datas: Any) {

    }

    override fun onLoading() {

    }
}

